### DISCUSSION TITLE
Include relationships when versioning with a history table

### DISCUSSION THREAD

**Discussion by njoeres** [sqlalchemy/sqlalchemy#10926] [Usage Questions]
In a new project I'm working with SQLAlchemy (2.0.21) and I would like to have versioned tables in my database. I found the [versioning with a history table](https://docs.sqlalchemy.org/en/20/orm/examples.html#module-examples.versioned_history) example in the docs, which is basically what I need...I guess^^

However, it seems that the relationships are not created for the version-tables. Below I tried to give an example of what my problem is.

```python
from examples.versioned_history.history_meta import Versioned
from examples.versioned_history.history_meta import versioned_session

from sqlalchemy import ForeignKey
from sqlalchemy import create_engine
from sqlalchemy.orm import Mapped
from sqlalchemy.orm import mapped_column
from sqlalchemy.orm import relationship
from sqlalchemy.orm import DeclarativeBase
from sqlalchemy.orm import declared_attr
from sqlalchemy.orm import sessionmaker
from typing import Optional


class Base(DeclarativeBase):
    id: Mapped[int] = mapped_column(primary_key=True)

class Parent(Versioned, Base):
    __tablename__ = "parent"
    name: Mapped[str]

    child: Mapped["Child"] = relationship(back_populates="parent")

class Child(Versioned, Base):
    __tablename__ = "child"
    value: Mapped[float]

    parent_id: Mapped[Optional[int]] = mapped_column(ForeignKey("parent.id"))
    parent: Mapped["Parent"] = relationship(back_populates="child")


engine = create_engine("sqlite:///:memory:", echo=False)

Session = sessionmaker(bind=engine)
versioned_session(Session)

Base.metadata.create_all(engine)
session = Session()

p = Parent(
    name="Parent",
    child=Child(value=1)
)

session.add(p)
session.commit()

assert p.version == 1

p.name = "Updated Parent"

session.commit()

assert p.version == 2

p.child = Child(value=2)

session.commit()
```

Until here everything works as expected. The following two things are not working.

## Problem 1
The following throws an AssertionError because `p.version` is 2, but I want `p.version` to be 3 because the child has changed, which is related to the parent.

```python
assert p.version == 3
```

## Problem 2
This does not work, as the HistoryClass doesn't have the same relations as its parent class. What I would like to have is a relation which tells me for this parent version, which was your child version.
```python
ParentHistory = Parent.__history_mapper__.class_
parent_history = session.query(ParentHistory).where(ParentHistory.id==p.id).first()
parent_history.child # A relationship to child does not exist for parent history
```

I was wondering how or if that would be possible? Maybe it is intended that this doesn't work because it's bad practice or there is a better way of how to achieve this?

In my project I have a lot of relationships, which are also nested and I would like to "pass through" the version to all related tables.

Many thanks in advance :)

**Comment by zzzeek** [comment]
the history example is focused on tracking changes to the underlying database table itself, not logical changes to objects.  If you want changes on the child object to do a version bump on the parent object, you can look into using [AttributeEvents](https://docs.sqlalchemy.org/en/20/orm/events.html#sqlalchemy.orm.AttributeEvents) to intercept those changes.   As far as relatiionships, you can add these to your History classes after the fact but I dont have a recipe for automating the production of copies of relatoinships.
